File: 110. Balanced Binary Tree.scala
/*
Given a binary tree, determine if it is height-balanced.

For this problem, a height-balanced binary tree is defined as a binary tree in which the depth of the two subtrees of every node never differ by more than 1.


*/

    1
   / \
  2   3

class TreeNode(var _value: Int) {
   var value: Int = _value
   var left: TreeNode = null
   var right: TreeNode = null
   override def toString = value+ ":"+ "{"+  left +","+right+"}"
}

def isBalanced(root: TreeNode): Boolean = {
    isBalancedImp(root)._1
}

def isBalancedImp(root:TreeNode): (Boolean,Int) = {
    if(root==null) return (true,0);
    val left = isBalancedImp(root.left);
    val right = isBalancedImp(root.right);
    if(left._1 && right._1 && math.abs(left._2 - right._2)<=1){
        return (true,math.max(left._2,right._2)+1);
    }else{
        return (false,math.max(left._2,right._2)+1);
    }
}

File: 110. Balanced Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.imp(root)[1]
    
    def imp(self,root):
        if root is None:
            return (0,True)
        else:
            left = self.imp(root.left)
            right = self.imp(root.right)
            if left[1] and right[1] and abs(left[0]-right[0])<2:
                return (1+max(right[0],left[0]),True)
            else :
                return  (1+max(right[0],left[0]),False)
